working/control-flow-collections/feature-specification.md

```diff
@@ -0,0 +1,913 @@
+      else
+        IconButton(icon: Icon(Icons.about)),
+    ]
+  );
+}
+```
+
+Users can and do use the conditional operator (`?:`) for cases like this today.
+It works OK, but isn't very easy on the eyes. And, of course, it falls down in
+cases where you don't have an "else" widget that you want to use instead of the
+"then" one.
+
+I admit it *is* a little strange seeing the familiar `if` keyword in a place
+where it's never appeared before. But my hope is that the semantics are fairly
+intuitive.
+
+### Repetition
+
+This is less common than conditional control flow, but repetition comes up too.
+At the statement level, you can loop if you want to execute something a certain
+number of times or for each of a series of items in an Iterable. In an
+expression context, it's useful if you want to produce more than one value.
+
+If we add spread syntax, then that covers some of these use cases. Spreading is
+fine, but when you want to do more than just insert a sequence in place, it
+forces you to chain a series of higher-order methods together to express what
+you want. That can get cumbersome, especially if you're mixing both repetition
+and conditional logic. You always *can* solve that using some combination of
+`map()`, `where()`, and `expand()`, but the result isn't always readable.
+
+So this proposal also lets you use `for` inside a collection literal. That
+turns, for example, this code:
+
+```dart
+var command = [
+  engineDartPath,
+  frontendServer,
+];
+for (var root in fileSystemRoots) {
+  command.add('--filesystem-root=$root');
+}
+for (var entryPointsJson in entryPointsJsonFiles) {
+  if (fileExists("$entryPointsJson.json")) {
+    command.add(entryPointsJson);
+  }
+}
+command.add(mainPath);
+```
+
+Into:
+
+```dart
+var command = [
+  engineDartPath,
+  frontendServer,
+  for (var root in fileSystemRoots) '--filesystem-root=$root'
+  for (var entryPointsJson in entryPointsJsonFiles)
+    if (fileExists("$entryPointsJson.json")) entryPointsJson,
+  mainPath
+];
+```
+
+Note the `if` nested inside the `for` and consider what that would look like if
+using higher-order methods on Iterable instead.
+
+A nice bonus of allowing `for` is that it gives us something not too far from
+the "[list comprehension][]" syntax supported by some other languages. We now
+have a nice short syntax for creating a list from a computation:
+
+[list comprehension]: https://en.wikipedia.org/wiki/List_comprehension
+
+```dart
+var integers = [for (var i = 1; i < 5; i++) i]; // [1, 2, 3, 4]
+var squares = [for (var n in integers) n * n]; // [1, 4, 9, 16]
+```
+
+It may seem surprising, but `for` also works perfectly well for map literals.
+It lets you turn this:
+
+```dart
+Map<String, WidgetBuilder>.fromIterable(
+  kAllGalleryDemos,
+  key: (demo) => '${demo.routeName}',
+  value: (demo) => demo.buildRoute,
+);
+```
+
+Into:
+
+```dart
+return {
+  for (var demo in kAllGalleryDemos)
+    '${demo.routeName}': demo.buildRoute,
```

Comment:
That colon there is suspiciously reminiscent of an argument list, eh?

```dart
  return Row(
    children: [
      IconButton(icon: Icon(Icons.menu)),
      Expanded(child: title),
      IconButton(
          icon: Icon(Icons.search),
          if (hasOnPressed) onPressed: onPressed),
    ]
  );
```

(Would be part of a separate proposal, hope I'm not spoiling the punchline!)

Reply:
See the last section of the doc. :)